python/whylogs/experimental/api/logger/__init__.py

```diff
@@ -76,41 +95,43 @@ def get_top_rank(row):
             ki_dict.columns = ["p@" + str(ki)]
         else:
             ki_dict["p@" + str(ki)] = ki_result
-
     output_data["average_precision" + ("_k_" + str(k) if k else "")] = ki_dict.mean(axis=1)
 
-    def _convert_non_numeric(row_dict):
-        return (
-            [
-                row_dict[target_column].index(pred_val) if pred_val in row_dict[target_column] else -1
-                for pred_val in row_dict[prediction_column]
-            ],
-            list(range(len(row_dict[prediction_column])))[::-1],
-        )
+    def _calc_non_numeric_relevance(row_dict):
+        prediction_relevance = []
+        ideal_relevance = []
+        for target_val in row_dict[prediction_column]:
+            ideal_relevance.append(1 if target_val in row_dict[target_column] else 0)
+            prediction_relevance.append(1 if target_val in row_dict[target_column] else 0)
+        for target_val in row_dict[target_column]:
+            if target_val not in row_dict[prediction_column]:
+                ideal_relevance.append(1)
+        return (prediction_relevance, ideal_relevance)
 
     if convert_non_numeric:
-        formatted_data[[prediction_column, target_column]] = formatted_data.apply(
-            _convert_non_numeric, result_type="expand", axis=1
+        formatted_data[["predicted_relevance", "ideal_relevance"]] = formatted_data.apply(
+            _calc_non_numeric_relevance, result_type="expand", axis=1
         )
+    else:
+        formatted_data["predicted_relevance"] = formatted_data[prediction_column]
+        formatted_data["ideal_relevance"] = formatted_data[target_column]
 
     def _calculate_row_ndcg(row_dict, k):
-        predicted_order = np.array(row_dict[prediction_column]).argsort()[::-1]
-        target_order = np.array(row_dict[target_column]).argsort()[::-1]
-        dcg_vals = [
-            (rel / math.log(i + 2, 2)) for i, rel in enumerate(np.array(row_dict[target_column])[predicted_order][:k])
-        ]
-        idcg_vals = [
-            (rel / math.log(i + 2, 2)) for i, rel in enumerate(np.array(row_dict[target_column])[target_order][:k])
-        ]
+        predicted_relevances = row_dict["predicted_relevance"]
+        ideal_relevances = sorted(row_dict["ideal_relevance"], reverse=True)
+        dcg_vals = [(rel / math.log(i + 2, 2)) for i, rel in enumerate(predicted_relevances[:k])]
+        idcg_vals = [(rel / math.log(i + 2, 2)) for i, rel in enumerate(ideal_relevances[:k])]
+        if sum(idcg_vals) == 0:
+            return 1  # if there is no relevant data, not much the recommender can do
```

Comment:
returning 1 is surprising from a ranking metrics perspective, for monitoring this might make some sense, but we will need to document the convention here.

In a follow up we might consider counting the number of rows this occurs, as it could indicate a problem in the data similar to counting nulls.